### ISSUE TITLE
Issues with AAAA handling

### ISSUE THREAD

**Issue by cbuben** [mesosphere/mesos-dns#102]
@kozyraki 

TL;DR: I think https://github.com/mesosphere/mesos-dns/pull/95 may have an issue.  This was a fix that added AAAA handling for mesos-managed zones.  However, in AAAA answers, mesos-dns responds that it is not authoritative for the zone and that recursion is not available.  This seems to blow up clients using newer versions of the glibc resolver.  I don't know why this is the case.

This behavior seems to be a product of [this](https://github.com/mesosphere/mesos-dns/commit/7dd99215a42f87664e6a37fd888ae34f28dfbb36#diff-c6a2b1bb9d7feeb56f43abf7caadeed1R297), where a new message is allocated, rather than using the message allocated [here](https://github.com/mesosphere/mesos-dns/blob/7dd99215a42f87664e6a37fd888ae34f28dfbb36/resolver/resolver.go#L228-L230), at the beginning of HandleMesos and is set as authoritative and supporting recursion.

When looking at authoritative AAAA answers that contain no records in other contexts outside of mesos-dns (i.e. analogous to the answers one would expect mesos-dns to provide), the answers seem to have both these flags set true.

If I tweak mesos-dns and reuse the message originally allocated at the beginning of HandleMesos, this issue does not occur (PR incoming).

Details
=====

docker 1.3.2 on CentOS 7, ipv6 disabled on host, mesos-dns https://github.com/mesosphere/mesos-dns/commit/f37051fc5a723eb021797e7d57d92755c011a28e.

Issue: clients of mesos-dns using (some) later versions of glibc resolver that send both A and AAAA questions for records in the mesos-dns owned zone experience resolution failures even when valid A answers are returned.  From what I can tell, CentOS 6 and Ubuntu 12.04 are unaffected, CentOS 7 and Ubuntu 14.04 exhibit the problem.  I'm not going to pretend I know the resolver changes that went into glibc, whether this behavior should be expected, etc.

In the failing cases, the AAAA answers (now handled via https://github.com/mesosphere/mesos-dns/pull/95) appear to cause the overall lookup operation to be unsuccessful from the perspective of clients using glibc resolver.  Lookups with dig and host work perfectly.  ping works perfectly as it appears to only ask for A records.  curl is a good test case to reproduce this problem.

In test cases, mesos-dns running locally on compute node, bound to 0.0.0.0/0, containers launched to point to address of docker0 bridge, and thus hit mesos-dns.

CentOS 7: affected
--------

curl without -4 flag:
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t centos:7 bash -c "curl -I http://master.mesos:5050"
curl: (6) Could not resolve host: master.mesos; No address associated with hostname
```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.62 -> 172.17.42.1  DNS 72 Standard query 0x24e3  A master.mesos
  2   0.000047  172.17.0.62 -> 172.17.42.1  DNS 72 Standard query 0xb4e7  AAAA master.mesos
  3   0.000394  172.17.42.1 -> 172.17.0.62  DNS 100 Standard query response 0x24e3  A 10.9.134.94
  4   0.000470  172.17.42.1 -> 172.17.0.62  DNS 72 Standard query response 0xb4e7 
  5   0.000556  172.17.0.62 -> 172.17.42.1  DNS 72 Standard query 0x461c  A master.mesos
  6   0.000586  172.17.0.62 -> 172.17.42.1  DNS 72 Standard query 0x0e5a  AAAA master.mesos
  7   0.000806  172.17.42.1 -> 172.17.0.62  DNS 100 Standard query response 0x461c  A 10.9.134.94
  8   0.000919  172.17.42.1 -> 172.17.0.62  DNS 72 Standard query response 0x0e5a 
```

curl with -4 flag
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t centos:7 bash -c "curl -4 -I http://master.mesos:5050"
HTTP/1.1 200 OK
Date: Mon, 16 Mar 2015 15:35:22 GMT
Content-Length: 4616
Content-Type: text/html

```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.63 -> 172.17.42.1  DNS 72 Standard query 0x43ab  A master.mesos
  2   0.000393  172.17.42.1 -> 172.17.0.63  DNS 100 Standard query response 0x43ab  A 10.9.134.94
```

Ubuntu 14.04: affected
--------

curl without -4 flag
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t ubuntu:14.04 bash -c "apt-get -qq install curl > /dev/null && curl -I http://master.mesos:5050"
curl: (6) Could not resolve host: master.mesos
```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.64 -> 172.17.42.1  DNS 78 Standard query 0x2131  A archive.ubuntu.com
  2   0.000039  172.17.0.64 -> 172.17.42.1  DNS 78 Standard query 0x978a  AAAA archive.ubuntu.com
  3   0.033661  172.17.42.1 -> 172.17.0.64  DNS 170 Standard query response 0x978a  AAAA 2001:67c:1360:8c01::19 AAAA 2001:67c:1360:8c01::18
  4   0.069854  172.17.42.1 -> 172.17.0.64  DNS 350 Standard query response 0x2131  A 91.189.91.23 A 91.189.91.24 A 91.189.92.200 A 91.189.92.201 A 91.189.88.149 A 91.189.91.13 A 91.189.91.14 A 91.189.91.15
4   5  21.476533  172.17.0.64 -> 172.17.42.1  DNS 72 Standard query 0x0e08  A master.mesos
  6  21.476579  172.17.0.64 -> 172.17.42.1  DNS 72 Standard query 0xdd27  AAAA master.mesos
  7  21.476916  172.17.42.1 -> 172.17.0.64  DNS 100 Standard query response 0x0e08  A 10.9.134.94
  8  21.476997  172.17.42.1 -> 172.17.0.64  DNS 72 Standard query response 0xdd27 
  9  21.477080  172.17.0.64 -> 172.17.42.1  DNS 72 Standard query 0x7994  A master.mesos
 10  21.477107  172.17.0.64 -> 172.17.42.1  DNS 72 Standard query 0xe406  AAAA master.mesos
 11  21.477347  172.17.42.1 -> 172.17.0.64  DNS 100 Standard query response 0x7994  A 10.9.134.94
 12  21.477402  172.17.42.1 -> 172.17.0.64  DNS 72 Standard query response 0xe406 
```

curl with -4 flag
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t ubuntu:14.04 bash -c "apt-get -qq install curl > /dev/null && curl -4 -I http://master.mesos:5050"
HTTP/1.1 200 OK
Date: Mon, 16 Mar 2015 15:52:13 GMT
Content-Length: 4616
Content-Type: text/html

```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.65 -> 172.17.42.1  DNS 78 Standard query 0x3e00  A archive.ubuntu.com
  2   0.000036  172.17.0.65 -> 172.17.42.1  DNS 78 Standard query 0x7619  AAAA archive.ubuntu.com
  3   0.069722  172.17.42.1 -> 172.17.0.65  DNS 170 Standard query response 0x7619  AAAA 2001:67c:1360:8c01::19 AAAA 2001:67c:1360:8c01::18
3   4   0.070809  172.17.42.1 -> 172.17.0.65  DNS 350 Standard query response 0x3e00  A 91.189.91.15 A 91.189.92.200 A 91.189.88.149 A 91.189.92.201 A 91.189.91.24 A 91.189.91.23 A 91.189.91.13 A 91.189.91.14
4   5  20.975650  172.17.0.65 -> 172.17.42.1  DNS 72 Standard query 0xb7ce  A master.mesos
  6  20.975862  172.17.42.1 -> 172.17.0.65  DNS 100 Standard query response 0xb7ce  A 10.9.134.94
```

CentOS 6: unaffected
--------

curl without -4 flag
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t centos:6 bash -c "curl -I http://master.mesos:5050"
HTTP/1.1 200 OK
Date: Mon, 16 Mar 2015 15:53:37 GMT
Content-Length: 4616
Content-Type: text/html


```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.66 -> 172.17.42.1  DNS 72 Standard query 0x4730  A master.mesos
  2   0.000056  172.17.0.66 -> 172.17.42.1  DNS 72 Standard query 0x6380  AAAA master.mesos
  3   0.000427  172.17.42.1 -> 172.17.0.66  DNS 100 Standard query response 0x4730  A 10.9.134.94
  4   0.000582  172.17.42.1 -> 172.17.0.66  DNS 72 Standard query response 0x6380 
```

Ubuntu 12.04: unaffected
--------
curl without -4 flag
```
[user@myhost ~]$ docker run --dns=172.17.42.1 -i -t ubuntu:12.04 bash -c "apt-get -qq install curl > /dev/null && curl -I http://master.mesos:5050"
debconf: delaying package configuration, since apt-utils is not installed
HTTP/1.1 200 OK
Date: Mon, 16 Mar 2015 15:55:59 GMT
Content-Length: 4616
Content-Type: text/html

```
tshark
```
[user@myhost ~]$ sudo tshark -s0 -ni docker0 udp and port 53 and host 172.17.42.1
Running as user "root" and group "root". This could be dangerous.
Capturing on 'docker0'
  1   0.000000  172.17.0.67 -> 172.17.42.1  DNS 78 Standard query 0x9c65  A archive.ubuntu.com
  2   0.070276  172.17.42.1 -> 172.17.0.67  DNS 350 Standard query response 0x9c65  A 91.189.91.24 A 91.189.91.23 A 91.189.92.200 A 91.189.91.14 A 91.189.91.15 A 91.189.88.149 A 91.189.91.13 A 91.189.92.201
  3   0.070602  172.17.0.67 -> 172.17.42.1  DNS 78 Standard query 0x3278  A archive.ubuntu.com
  4   0.105029  172.17.42.1 -> 172.17.0.67  DNS 350 Standard query response 0x3278  A 91.189.91.15 A 91.189.88.149 A 91.189.92.201 A 91.189.91.23 A 91.189.92.200 A 91.189.91.24 A 91.189.91.13 A 91.189.91.14
4   5  11.368244  172.17.0.67 -> 172.17.42.1  DNS 72 Standard query 0xafab  AAAA master.mesos
  6  11.368446  172.17.42.1 -> 172.17.0.67  DNS 72 Standard query response 0xafab 
  7  11.368506  172.17.0.67 -> 172.17.42.1  DNS 72 Standard query 0xafab  AAAA master.mesos
  8  11.368805  172.17.42.1 -> 172.17.0.67  DNS 72 Standard query response 0xafab 
  9  11.368900  172.17.0.67 -> 172.17.42.1  DNS 72 Standard query 0xc206  AAAA master.mesos
 10  11.369114  172.17.42.1 -> 172.17.0.67  DNS 72 Standard query response 0xc206 
 11  11.369187  172.17.0.67 -> 172.17.42.1  DNS 72 Standard query 0xc206  AAAA master.mesos
 12  11.369475  172.17.42.1 -> 172.17.0.67  DNS 72 Standard query response 0xc206 
 13  11.369560  172.17.0.67 -> 172.17.42.1  DNS 72 Standard query 0xb5b5  A master.mesos
 14  11.369864  172.17.42.1 -> 172.17.0.67  DNS 100 Standard query response 0xb5b5  A 10.9.134.94
```

AAAA answer
---------

Here's an example of the AAAA answer with a full tshark decode of a failed request from a CentOS 7 client, notice non-authoritative and no recursion...
```
Internet Protocol Version 4, Src: 172.17.42.1 (172.17.42.1), Dst: 172.17.0.70 (172.17.0.70)
    Version: 4
    Header length: 20 bytes
    Differentiated Services Field: 0x00 (DSCP 0x00: Default; ECN: 0x00: Not-ECT (Not ECN-Capable Transport))
        0000 00.. = Differentiated Services Codepoint: Default (0x00)
        .... ..00 = Explicit Congestion Notification: Not-ECT (Not ECN-Capable Transport) (0x00)
    Total Length: 58
    Identification: 0x95c0 (38336)
    Flags: 0x02 (Don't Fragment)
        0... .... = Reserved bit: Not set
        .1.. .... = Don't fragment: Set
        ..0. .... = More fragments: Not set
    Fragment offset: 0
    Time to live: 64
    Protocol: UDP (17)
    Header checksum: 0x2289 [correct]
        [Good: True]
        [Bad: False]
    Source: 172.17.42.1 (172.17.42.1)
    Destination: 172.17.0.70 (172.17.0.70)
User Datagram Protocol, Src Port: 53 (53), Dst Port: 39289 (39289)
    Source port: 53 (53)
    Destination port: 39289 (39289)
    Length: 38
    Checksum: 0x82a1 [validation disabled]
        [Good Checksum: False]
        [Bad Checksum: False]
Domain Name System (response)
    [Request In: 6]
    [Time: 0.000436000 seconds]
    Transaction ID: 0x4940
    Flags: 0x8100 Standard query response, No error
        1... .... .... .... = Response: Message is a response
        .000 0... .... .... = Opcode: Standard query (0)
        .... .0.. .... .... = Authoritative: Server is not an authority for domain
        .... ..0. .... .... = Truncated: Message is not truncated
        .... ...1 .... .... = Recursion desired: Do query recursively
        .... .... 0... .... = Recursion available: Server can't do recursive queries
        .... .... .0.. .... = Z: reserved (0)
        .... .... ..0. .... = Answer authenticated: Answer/authority portion was not authenticated by the server
        .... .... ...0 .... = Non-authenticated data: Unacceptable
        .... .... .... 0000 = Reply code: No error (0)
    Questions: 1
    Answer RRs: 0
    Authority RRs: 0
    Additional RRs: 0
    Queries
        master.mesos: type AAAA, class IN
            Name: master.mesos
            Type: AAAA (IPv6 address)
            Class: IN (0x0001)
```

**Comment by kozyraki** [comment]
@cbuben thanks a lot for this. I will look into it later today.

**Comment by kozyraki** [comment]
@cbuben Thanks you SO MUCH. While testing another PR, I run into this issue. It would probably take me quite some time to figure it out. I owe you one.

**Comment by imrangit** [comment]
I am not sure it has resolved for CentOS 6.5 as well. I am having issues with the failed lookups. I am using mesos-dns 0.1.2 on CentOS 6.5. My lookups fail with "host" command whereas it works with "dig".

[ishaikh@mesosdev4 ~]$ host esdemo-1325-s4.elasticsearch.mesos
esdemo-1325-s4.elasticsearch.mesos has address 10.1.160.30
Host esdemo-1325-s4.elasticsearch.mesos not found: 3(NXDOMAIN)

[ishaikh@mesosdev4 ~]$ dig +short esdemo-1325-s4.elasticsearch.mesos
10.1.160.30

Please advise.

**Comment by kozyraki** [comment]
If I remember correctly, host will query ALL the available name servers.  You seem. Have two in you etc/resolv.conf. One is mesos-DNS and it provides a correct result, the second one cannot find mesos tasks. Dig only queries a single name server, the first one on you config file.

**Comment by kamoljan** [comment]
Hi @kozyraki 

I am not sure if this issue is solved already.
But we are getting exactly the same symptoms.

In marathon app we are accessing DNS (e.g. api.xxx.com) from DCOS cluster. 
In heavy load, we are getting sometimes `curl: (6) Could not resolve host: api.xxx.com` from:
```
curl api.xxx.com 
```
However, with 
```
curl api.parse.com -4
```
it works.

Is it something to do with our `resolver` or Mesos-DNS?

Our resolver:
```
$ cat /etc/resolv.conf
# Generated by gen_resolvconf.py. Do not edit.
# Change configuration options by changing DC/OS cluster configuration.
# This file must be overwritten regularly for proper cluster operation around
# master failure.

options timeout:1
options attempts:3

nameserver 198.51.100.1
nameserver 198.51.100.2
nameserver 198.51.100.3
```

All nameservers are OK.

EC2 Instance type:
```
$ uname -a
Linux ip-10-0-4-229.ap-southeast-1.compute.internal 4.6.3-coreos #2 SMP Tue Aug 23 02:14:16 UTC 2016 x86_64 Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz GenuineIntel GNU/Linux
```

Please advice
